ocaml/typing/typedecl.ml

```diff
@@ -1113,40 +1171,119 @@ let update_decl_jkind env dpath decl =
     | _, Record_boxed jkinds ->
       let lbls, all_void = update_label_jkinds env loc lbls (Some jkinds) in
       let jkind = Jkind.for_boxed_record ~all_void in
-      let has_values, has_floats =
-        Array.fold_left
-          (fun (values, floats) jkind ->
-             match Jkind.get_default_value jkind with
-             | Value | Immediate64 | Immediate -> (Has_values, floats)
-             | Float64 -> (values, Has_float64s)
-             | Void -> (values, floats)
-             (* CR layouts v2.1: make unboxed ints work with records *)
-             | Word | Bits32 | Bits64 ->
-              Misc.fatal_error "Typedecl.update_record_kind: no support for unboxed ints"
-             | Any -> assert false)
-          (No_values, No_float64s) jkinds
+      let classify (lbl : Types.label_declaration) jkind =
+        if is_float env lbl.ld_type then Float_element
+        else match Jkind.get_default_value jkind with
+          | Value | Immediate64 -> Value_element
+          | Immediate -> Flat_imm_element
+          | Float64 -> Flat_float64_element
+          | Void -> Element_without_runtime_component
+          | Word | Bits32 | Bits64 ->
+              Misc.fatal_error
+                "Typedecl.update_record_kind: no support for unboxed ints"
+          | Any ->
+              Misc.fatal_error "Typedecl.update_record_kind: unexpected Any"
+      in
+      let classifications =
+        List.mapi (fun i lbl -> lbl, classify lbl jkinds.(i)) lbls
       in
+      let element_reprs =
+        { values = false; imms = false; floats = false; float64s = false }
+      in
+      List.iter
+        (fun (_lbl, repr) ->
+           match repr with
+           | Float_element -> element_reprs.floats <- true
+           | Flat_imm_element -> element_reprs.imms <- true
+           | Flat_float64_element -> element_reprs.float64s <- true
+           | Value_element -> element_reprs.values <- true
+           | Element_without_runtime_component -> ())
+        classifications;
       let rep =
```

Comment:
I did the naive refactor (pull out the function and turn its free variables into parameters) and I don't think it's more readable. I did a smaller refactor (just pull out `find_flat_suffix`) and think that's much more readable. So that's what I did. Thanks for the suggestion.